iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp

```diff
@@ -0,0 +1,247 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===- BufferizeCopyOnlyDispatchesPassPass.cpp ----------------------------===//
+//
+// This pass converts dispatches that are copy only into a form where backends
+// can tile and distribute them appropriately.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Interfaces/ViewLikeInterface.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Adds to `OpFoldResult`s and returns the result as an `OpFoldResult`.
+static OpFoldResult add(OpBuilder &builder, Location loc, OpFoldResult lhs,
+                        OpFoldResult rhs) {
+  auto lhsAttr = lhs.dyn_cast<Attribute>();
+  auto rhsAttr = rhs.dyn_cast<Attribute>();
+  if (lhsAttr && rhsAttr) {
+    int64_t result = lhsAttr.cast<IntegerAttr>().getInt() +
+                     rhsAttr.cast<IntegerAttr>().getInt();
+    return builder.getIndexAttr(result);
+  }
+  // Generate the affine.apply that computes the result
+  SmallVector<Value> operands;
+  AffineExpr resultExpr = nullptr;
+  auto addToResult = [&](OpFoldResult ofr) {
+    AffineExpr e;
+    if (auto attr = ofr.dyn_cast<Attribute>()) {
+      e = getAffineConstantExpr(attr.cast<IntegerAttr>().getInt(),
+                                builder.getContext());
+    } else {
+      e = getAffineSymbolExpr(operands.size(), builder.getContext());
+      operands.push_back(ofr.get<Value>());
+    }
+    resultExpr = resultExpr ? resultExpr + e : e;
+  };
+  addToResult(lhs);
+  addToResult(rhs);
+  AffineMap map = AffineMap::get(0, operands.size(), resultExpr);
+  return builder.create<AffineApplyOp>(loc, map, operands).getResult();
+}
+
+/// Returns the offsets to use when combining two operations that implement the
+/// `OffsetSizeAndStrideOpInterface`. Also checks that the strides are 1.
+static FailureOr<SmallVector<OpFoldResult>> foldOffsetsSizesAndStrides(
```

Comment:
Maybe name it as `composeOffsets`?

Reply:
I actually updated to handle non-unit strides cause it was needed. So keeping the name for now.